File: SharedUtils/src/test/java/com/braintreepayments/api/sharedutils/HttpClientUnitTest.java
package com.braintreepayments.api.sharedutils;

import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoInteractions;
import static org.mockito.Mockito.when;

import org.junit.Before;
import org.junit.Test;

public class HttpClientUnitTest {

    private SynchronousHttpClient syncHttpClient;

    private HttpRequest httpRequest;
    private MockThreadScheduler threadScheduler;

    @Before
    public void beforeEach() {
        syncHttpClient = mock(SynchronousHttpClient.class);
        threadScheduler = spy(new MockThreadScheduler());

        httpRequest = new HttpRequest().path("https://example.com");
    }

    @Test
    public void sendRequest_sendsRequestOnBackgroundThread() throws Exception {
        HttpClient sut = new HttpClient(syncHttpClient, threadScheduler);

        NetworkResponseCallback callback = mock(NetworkResponseCallback.class);
        sut.sendRequest(httpRequest, callback);

        verifyNoInteractions(syncHttpClient);
        threadScheduler.flushBackgroundThread();

        verify(syncHttpClient).request(httpRequest);
    }

    @Test
    public void sendRequest_whenBaseHttpClientThrowsException_notifiesErrorViaCallbackOnMainThread()
            throws Exception {
        HttpClient sut = new HttpClient(syncHttpClient, threadScheduler);

        Exception exception = new Exception("error");
        when(syncHttpClient.request(httpRequest)).thenThrow(exception);

        NetworkResponseCallback callback = mock(NetworkResponseCallback.class);
        sut.sendRequest(httpRequest, callback);

        threadScheduler.flushBackgroundThread();
        verify(callback, never()).onResult(null, exception);

        threadScheduler.flushMainThread();
        verify(callback).onResult(null, exception);
    }

    @Test
    public void sendRequest_onBaseHttpClientRequestSuccess_notifiesSuccessViaCallbackOnMainThread()
            throws Exception {
        HttpClient sut = new HttpClient(syncHttpClient, threadScheduler);
        HttpResponse response = new HttpResponse("response body", new HttpResponseTiming(123, 456));

        when(syncHttpClient.request(httpRequest)).thenReturn(response);

        NetworkResponseCallback callback = mock(NetworkResponseCallback.class);
        sut.sendRequest(httpRequest, callback);

        threadScheduler.flushBackgroundThread();
        verify(callback, never()).onResult(response, null);

        threadScheduler.flushMainThread();
        verify(callback).onResult(response, null);
    }

    @Test
    public void sendRequest_whenCallbackIsNull_doesNotNotifySuccess() throws Exception {
        HttpClient sut = new HttpClient(syncHttpClient, threadScheduler);
        HttpResponse response = new HttpResponse("response body", new HttpResponseTiming(123, 456));

        when(syncHttpClient.request(httpRequest)).thenReturn(response);
        sut.sendRequest(httpRequest, null);

        threadScheduler.flushBackgroundThread();
        verify(threadScheduler, never()).runOnMain(any(Runnable.class));
    }

    @Test
    public void sendRequest_whenCallbackIsNull_doesNotNotifyError() throws Exception {
        HttpClient sut = new HttpClient(syncHttpClient, threadScheduler);

        Exception exception = new Exception("error");
        when(syncHttpClient.request(httpRequest)).thenThrow(exception);

        sut.sendRequest(httpRequest, null);

        threadScheduler.flushBackgroundThread();
        verify(threadScheduler, never()).runOnMain(any(Runnable.class));
    }

    @Test
    public void sendRequestSynchronous_sendsHttpRequest() throws Exception {
        HttpClient sut = new HttpClient(syncHttpClient, threadScheduler);
        HttpResponse response = new HttpResponse("response body", new HttpResponseTiming(123, 456));

        when(syncHttpClient.request(httpRequest)).thenReturn(response);

        String result = sut.sendRequest(httpRequest);
        assertEquals("response body", result);
    }
}

File: SharedUtils/src/test/java/com/braintreepayments/api/sharedutils/HttpClientUnitTest.kt
package com.braintreepayments.api.sharedutils

import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.StandardTestDispatcher
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import kotlin.test.assertFailsWith

@OptIn(ExperimentalCoroutinesApi::class)
@RunWith(RobolectricTestRunner::class)
class HttpClientUnitTest {

    private val testDispatcher = StandardTestDispatcher()
    private lateinit var mockOkHttpClient: OkHttpSynchronousHttpClient
    private lateinit var sut: HttpClient


    @Before
    fun setUp() {
        mockOkHttpClient = mockk<OkHttpSynchronousHttpClient>()
        sut = HttpClient(mockOkHttpClient, testDispatcher)
    }

    @Test
    fun `sendRequest is a suspend function that executes in a coroutine`() =
        runTest(testDispatcher) {
        val request = mockk<OkHttpRequest>()
        val mockResponse = mockk<HttpResponse>()

        every { mockOkHttpClient.executeRequest(request) } returns mockResponse

        // if sendRequest is not a suspend function it wouldn't be able to run within runTest coroutine
        val result = sut.sendRequest(request)

        assertEquals(mockResponse, result)
        verify { mockOkHttpClient.executeRequest(request) }
    }

    @Test
    fun `when sendRequest is called with successful response, returns response`() =
        runTest(testDispatcher) {
        val request = mockk<OkHttpRequest>()
        val mockResponse = mockk<HttpResponse>()

        every { mockOkHttpClient.executeRequest(request) } returns mockResponse

        val result = sut.sendRequest(request)

        assertEquals(mockResponse, result)
        verify { mockOkHttpClient.executeRequest(request) }
    }

    @Test
    fun `when sendRequest is called with exception, throws exception`() =
        runTest(testDispatcher) {
        val request = mockk<OkHttpRequest>()
        val exception = RuntimeException("Network error")

        every { mockOkHttpClient.executeRequest(request) } throws exception

        val thrownException = assertFailsWith<RuntimeException> {
            sut.sendRequest(request)
        }

        assertEquals("Network error", thrownException.message)
        verify { mockOkHttpClient.executeRequest(request) }
    }

    @Test
    fun `when IOException occurs, throws IOException`() =
        runTest(testDispatcher) {
        val request = mockk<OkHttpRequest>()
        val ioException = java.io.IOException("Network timeout")

        every { mockOkHttpClient.executeRequest(request) } throws ioException

        val thrownException = assertFailsWith<java.io.IOException> {
            sut.sendRequest(request)
        }

        assertEquals("Network timeout", thrownException.message)
        verify { mockOkHttpClient.executeRequest(request) }
    }

    @Test
    fun `when HttpClient constructor is called without parameters, default instances are created`() {
        val sut = HttpClient()
        assertNotNull(sut)
    }
}
